library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -124,18 +105,16 @@ impl<T: Clone, A: Allocator + Clone> Clone for VecDeque<T, A> {
     }
 
     fn clone_from(&mut self, other: &Self) {
-        self.truncate(other.len());
-
-        let mut iter = PairSlices::from(self, other);
-        while let Some((dst, src)) = iter.next() {
-            dst.clone_from_slice(&src);
-        }
+        self.clear();
+        self.head = 0;
+        self.reserve(other.len);
 
-        if iter.has_remainder() {
-            for remainder in iter.remainder() {
-                self.extend(remainder.iter().cloned());
-            }
+        let (a, b) = other.as_slices();
+        unsafe {
+            self.write_iter(0, a.iter().cloned(), &mut 0);
+            self.write_iter(a.len(), b.iter().cloned(), &mut 0);
         }
+        self.len = other.len;
```

Comment:
setting `len` like this makes me nervous.

It's either unnecessary, because `write_iter` already did the right thing, or this is leaking any already-cloned items if one of the clone calls failed.

Please comment it to talk about panic safety for this case.

Reply:
Yes this is probably not the best way to do it. `write_iter` itself doesn't update the length on its own. `write_iter_wrapping` does, it uses a guard type to update the deques length in the Drop impl. Either way, it'd probably be best to just do `self.extend(other.iter().cloned())`.